### ISSUE TITLE
sql: unexpectedly high memory usage when sorting/joining wide rows

### ISSUE THREAD

**Issue by mlazowik** [cockroachdb/cockroach#68008]
When sorting/joining wide rows (especially when spilling to disk) memory usage can get way higher than workmem, up to gigabytes per single query.

# To Reproduce

## sort

```sql
create table t (id primary key, t, v)
as
select gen_random_uuid(), now(), '{"value": "' || repeat('hi', 4 * 64 * 1024) || '"}'
from generate_series(1,1000) g(g);

analyze t;

explain analyze (distsql)
select * from t order by t desc;
```

This results in `max memory allocated: 1.5 GiB`. Notice that scratch disk usage is very low compared to that (24 MiB). Interestingly, if you query before statistics get calculated, the query is slower, but also uses less memory. (This is why we do `analyze t` in the repro)

The memory usage seems to scale linearly with the row width. With `repeat('hi', 64 * 1024 * 20)` you can even get `ERROR: internal error: unexpected error from the vectorized engine: runtime error: slice bounds out of range [:-2145375836]` (we're not interested in _that_ wide rows tho.)

## join

```sql
create table t1 (id primary key, n, v)
as
select gen_random_uuid(), 1, '{"value": "' || repeat('hi', 4 * 64 * 1024) || '"}'
from generate_series(1,1000) g(g);

create table t2 (n) as values (1);

analyze t1;

explain analyze (distsql)
select * from t1 inner join t2 on t1.n = t2.n;
```

This results in 518 MiB sql memory used.

# Expected behavior
We'd like memory usage to stay on the order of workmem.

# Environment
 - CockroachDB version 21.1.6
 - Server OS: linux/macOS
 - Client app `cockroach sql`

**Comment by yuzefovich** [comment]
Thanks @mlazowik for a very detailed report.

Essentially you're running into the worst case behavior here. The problem is that currently we don't know the average size of rows that we're about to scan, so we're taking a guess - for `Bytes`-like types we guess that each column is 64 bytes in size. As a result, a single row in table `t` is estimated to take up on the order of 100 bytes. `workmem` limit is 64MiB, so we think that we can use columnar batches of maximum capacity (1024). As a result, the scan operator allocates the batch with 1000  rows (based on the table size hint), and we're performing the full table scan. While we're doing it, we currently don't look the actual row sizes after we have scanned them.

Such a behavior will scale linearly with increasing the width of rows; however, once the number of rows exceeds 1024, we will be reusing the previously allocated memory, so the memory usage growth will stop.

In regards to the index out of bounds - this is also a known limitation (due to the usage of `int32` to store the offsets for bytes-like types).

I don't have a proposal or a workaround at the moment, we'll need to think a bit about what to do. But generally speaking, Cockroach doesn't like wide rows, and this is one example.

**Comment by mlazowik** [comment]
Thank you for the explanation!

> Once the number of rows exceeds 1024, we will be reusing the previously allocated memory, so the memory usage growth will stop.

Yeah, we've seen it taper off, but weren't sure how much we could depend on that. We were mostly worried this would go up with number of rows always, if that's not the case we can sort of offset the issue by https://downloadmoreram.com

Our schema is mostly regular stuff + one/two large JSON columns, we only store and load the data from them, no joins/filters/anything. I guess we can either think how to move them off the db, or, hackier but maybe easier – split those columns into ~1000 text/bytes columns and split/stitch app-side?

**Comment by mlazowik** [comment]
> Once the number of rows exceeds 1024, we will be reusing the previously allocated memory, so the memory usage growth will stop.

Hm, on our prod we were over 1024 rows from the start and we had to _increase_ workmem as more rows came in to stop hitting sql mem limit. I'll look into creating repros for higher mem usage with lower workmem.

**Comment by jordanlewis** [comment]
This would be improved (and possibly solved) by #55697, which would allow us to properly size the vectorized batches ahead of time.

**Comment by yuzefovich** [comment]
There is a bit more nuance in how we're accounting for the memory used. The `scan` operator will be reusing the same `coldata.Batch` for all rows read (unless they differ in size and force increasing the initially allocated space), and this batch should always take up on the order of 1000 x 512 KiB = 500MiB (given `repeat('hi', 4 * 64 * 1024)`). Then, whenever we're performing a buffering operation (like a sort or a hash join) we will need to allocate a separate `coldata.Batch` that will be accounted for separately. So let's take the sort as an example - the sort copies over the input rows into a buffer (until `workmem` limit), and if the limit is exceeded, then it'll spill to disk.

In the example with 1000 rows, the sort will copy the input batch into a buffer, perform the memory accounting, will see that `workmem` limit is exceeded, and spills to disk. We're probably still account for this copy to the buffer (I'm not certain - I'd have expected this copy to not be reflected in the memory accounting system, but otherwise the math doesn't add up). Spilling to disk occurs - which means that we're exporting the buffer and writing all the data to disk (note that we compress the data when spilling, that's why usually it'll be significantly smaller), and the external sort will only then realize that it has actually consumed the input fully (by asking for more rows from the scan and receiving none). We're performing the external merge sort on a batch granularity, so we won't split the 500MiB batch and will actually sort the whole 500MiB batch in memory, and we will be accounting for it too. Thus we see we have 3 places reporting 500MiB each.

In regards to the table statistics being important. Without the knowledge of the table size, the `coldata.Batch` size in the `scan` operator will grow exponentially starting from 1 (i.e. 1, 2, 4, 8, ...). This way we allocate batch of small capacity only to discard it on the next iteration, and 1000 rows is actually about the worst possible for this pattern since we never reuse memory. However, it will report less memory because the batch of maximum number of rows will be 512 (not 1000), we will also fallback to the external sorter sooner, and because we have more batches, we will be more efficient with the external merge sort.

As Jordan pointed out, #55697 will significantly help us with the initial estimates, and I believe it is on our medium term roadmap.

**Comment by DrewKimball** [comment]
@yuzefovich do you think it would be a good idea to keep track of the output batch size in `cfetcher`, and emit immediately once the memory limit is reached? (Like we already do for limit hints)

**Comment by yuzefovich** [comment]
I was thinking about it - we definitely have a convenient place to do that after every row is finalized, but I was afraid of the performance hit of measuring the memory footprint for each row. However, if we do something smart/optimized as we did for `PerformAppend`, the perf hit might be negligible.

**Comment by DrewKimball** [comment]
Maybe we could add up the raw key value bytes - it wouldn't be exact, but might be good enough to prevent situations like this one.

**Comment by yuzefovich** [comment]
Hey @mlazowik, just wanted to let you know that we have merged some fixes already with one more to go which significantly improved the situation. The sort query now reports the usage of 209MiB whereas the hash joiner one of 139MiB.

**Comment by mlazowik** [comment]
Well that fast fast, thank you all!

**Comment by mlazowik** [comment]
Any chance of backporting these to v21.1? I'm guessing v21.2 will come out in around 2 months?

**Comment by yuzefovich** [comment]
We probably won't backport the fixes to 21.1 since they are quite involved, but you're correct that 21.2 should be out in 2-3 months.
